pkg/util/http/transport.go

```diff
@@ -57,46 +68,75 @@ func CreateHTTPTransport() *http.Transport {
-		// check no_proxy list first
-		for _, host := range p.NoProxy {
-			if r.URL.Host == host {
-				log.Debugf("URL match no_proxy list item '%s': not using any proxy", host)
-				return nil, nil
-			}
-		}
 
-		// check proxy by scheme
-		confProxy := ""
-		if r.URL.Scheme == "http" {
-			confProxy = p.HTTP
-		} else if r.URL.Scheme == "https" {
-			confProxy = p.HTTPS
-		} else {
-			log.Warnf("Proxy configuration do not support scheme '%s'", r.URL.Scheme)
+	proxyConfig := &httpproxy.Config{
+		HTTPProxy:  p.HTTP,
+		HTTPSProxy: p.HTTPS,
+		NoProxy:    strings.Join(p.NoProxy, ","),
+	}
+
+	if config.Datadog.GetBool("proxy.no_proxy_nonexact_match") {
+		return func(r *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(r.URL)
 		}
+	}
 
-		if confProxy != "" {
-			proxyURL, err := url.Parse(confProxy)
-			if err != nil {
-				err := fmt.Errorf("Could not parse the proxy URL for scheme %s from configuration: %s", r.URL.Scheme, err)
-				log.Error(err.Error())
-				return nil, err
+	return func(r *http.Request) (*url.URL, error) {
+		url, err := func(r *http.Request) (*url.URL, error) {
+			// check no_proxy list first
+			for _, host := range p.NoProxy {
+				if r.URL.Host == host {
+					log.Debugf("URL match no_proxy list item '%s': not using any proxy", host)
+					return nil, nil
+				}
 			}
-			userInfo := ""
-			if proxyURL.User != nil {
-				if _, isSet := proxyURL.User.Password(); isSet {
-					userInfo = "*****:*****@"
-				} else {
-					userInfo = "*****@"
+
+			// check proxy by scheme
+			confProxy := ""
+			if r.URL.Scheme == "http" {
+				confProxy = p.HTTP
+			} else if r.URL.Scheme == "https" {
+				confProxy = p.HTTPS
+			} else {
+				log.Warnf("Proxy configuration do not support scheme '%s'", r.URL.Scheme)
+			}
+
+			if confProxy != "" {
+				proxyURL, err := url.Parse(confProxy)
+				if err != nil {
+					err := fmt.Errorf("Could not parse the proxy URL for scheme %s from configuration: %s", r.URL.Scheme, err)
+					log.Error(err.Error())
+					return nil, err
 				}
+				userInfo := ""
+				if proxyURL.User != nil {
+					if _, isSet := proxyURL.User.Password(); isSet {
+						userInfo = "*****:*****@"
+					} else {
+						userInfo = "*****@"
+					}
+				}
+
+				log.Debugf("Using proxy %s://%s%s for URL '%s'", proxyURL.Scheme, userInfo, proxyURL.Host, SanitizeURL(r.URL.String()))
+				return proxyURL, nil
 			}
 
-			log.Debugf("Using proxy %s://%s%s for URL '%s'", proxyURL.Scheme, userInfo, proxyURL.Host, SanitizeURL(r.URL.String()))
-			return proxyURL, nil
+			// no proxy set for this request
+			return nil, nil
+		}(r)
+
+		// Print a warning if the proxy behavior would change if the new no_proxy behavior would be enabled
+		newURL, _ := proxyConfig.ProxyFunc()(r.URL)
+		if url != newURL && url != nil {
+			urlString := r.URL.String()
+			NoProxyWarningMapMutex.Lock()
+			if _, ok := NoProxyWarningMap[urlString]; !ok {
+				NoProxyWarningMap[SanitizeURL(r.URL.String())] = true
```

Comment:
Could URL contains sensitive data (user/password, and argument). We don't want to those to be sent to our backend. We might want to only to use the scheme and domain.

Reply:
I corrected the log lines so they are safer as you suggested.